qualtran/bloqs/rotations/phase_gradient.py

```diff
@@ -214,46 +215,66 @@ class AddScaledValIntoPhaseReg(GateWithRegisters, cirq.ArithmeticGate):
         (https://arxiv.org/abs/2007.07391), Appendix A: Addition for controlled rotations
     """
 
-    inp_bitsize: int
+    inp_dtype: QFxp
     phase_bitsize: int
     gamma: float
     gamma_bitsize: int
 
-    def __attrs_post_init__(self):
-        assert 0 <= self.gamma <= 1
-
     @cached_property
     def signature(self):
-        return Signature.build_from_dtypes(
-            x=QUInt(self.inp_bitsize), phase_grad=QFxp(self.phase_bitsize, self.phase_bitsize)
-        )
+        return Signature.build_from_dtypes(x=self.inp_dtype, phase_grad=self.phase_dtype)
 
     def registers(self):
-        return [2] * self.inp_bitsize, [2] * self.phase_bitsize
+        return [2] * self.inp_dtype.num_qubits, [2] * self.phase_bitsize
 
     def with_registers(self, *new_registers: Union[int, Sequence[int]]):
         raise NotImplementedError("not needed.")
 
     @cached_property
-    def gamma_int_numerator(self) -> int:
-        _, gamma_fixed_width_int = float_as_fixed_width_int(self.gamma, self.gamma_bitsize + 1)
-        gamma_fixed_width_float = gamma_fixed_width_int / 2**self.gamma_bitsize
-        _, gamma_numerator = float_as_fixed_width_int(
-            gamma_fixed_width_float, self.phase_bitsize + 1
-        )
-        return gamma_numerator
+    def phase_dtype(self) -> QFxp:
+        return QFxp(self.phase_bitsize, self.phase_bitsize, signed=False)
+
+    @cached_property
+    def gamma_dtype(self) -> QFxp:
+        n_int = Fxp(abs(self.gamma), signed=False).n_int
+        return QFxp(n_int + self.gamma_bitsize, self.gamma_bitsize, signed=False)
+
+    @cached_method
+    def scaled_val(self, x: int) -> int:
+        sign = np.sign(self.gamma)
+        x_fxp = Fxp(x, dtype=self.inp_dtype.fxp_dtype_str)
+        gamma_fxp = Fxp(abs(self.gamma), dtype=self.gamma_dtype.fxp_dtype_str)
```

Comment:
A little comment here and what you're trying to achieve would help the reader :)

Reply:
added detailed docstrings